btf/ext_info.go

```diff
@@ -139,7 +139,15 @@ var nativeEncoderPool = sync.Pool{
 
 // MarshalExtInfos encodes function and line info embedded in insns into kernel
 // wire format.
+//
+// Returns ErrNotSupported if the kernel doesn't support BTF-associated programs.
 func MarshalExtInfos(insns asm.Instructions) (_ *Handle, funcInfos, lineInfos []byte, _ error) {
+	// Bail out early if the kernel doesn't support Func(Proto). If this is the
+	// case, func_info will also be unsupported.
+	if err := haveProgBTF(); err != nil {
```

Comment:
Here we ask for permission, but for maps we ask for forgiveness. Why not ask for forgiveness in both cases?

Benefit: errors from marshaling ext infos are surfaced on old kernels as well. Downside: we pay the cost on old kernels that don't have support.

Reply:
> Here we ask for permission, but for maps we ask for forgiveness. Why not ask for forgiveness in both cases?

This is intentional, and this question is part of what makes this issue so tricky.

- Traversing insns and marshaling a btf blob and extinfos is significantly more expensive than calling a (potentially cached) `haveProgBTF()`.
- Loading a func(proto) is a perfect oracle for both the kernel supporting the BTF kind, as well as func_infos. Unlike `haveMapBTF()`, which is still no guarantee that the map types supports BTF metadata.

> Benefit: errors from marshaling ext infos are surfaced on old kernels as well.

Not really. `haveProgBTF()` != nil means the kernel doesn't support func_infos (and, by proxy, line_infos), so there's no point in trying to marshal and load any of it. Short-circuiting here looks like the best approach to me.

Reply:
> extinfos is significantly more expensive than calling a (potentially cached) haveProgBTF()

True, however the happy path aka new kernels won't have that overhead.